Feature: Add edit set annotation

adberger opened the pull request:
This PR adds the `edit set annotation` to kustomize

Most code is copied from `kustomize/commands/edit/set/setlabel.go` as suggested.
Improved testing of annotation key validity as described in https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/

resolves #4051

natasha41575 commented:
/ok-tot-test

natasha41575 commented:
ugh, I can't spell
/ok-to-test

natasha41575 commented on kustomize/commands/edit/set/setannotation.go:
```diff
@@ -0,0 +1,99 @@
+		return err
+	}
+	for key := range m {
+		if IsValidKey(key) == false {
```
```suggestion
		if !IsValidKey(key) {
```

natasha41575 commented on kustomize/commands/edit/set/setannotation.go:
```diff
@@ -0,0 +1,99 @@
+	mapValidator func(map[string]string) error
+}
+
+// first part for prefix segment (DNS1123Label) of an annotation followed by a slash, second part for name segment of an annotation
```
please add a link to the documentation for valid annotation keys, e.g.
```
// see https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/
```

natasha41575 commented on kustomize/commands/edit/set/setannotation.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+// first part for prefix segment (DNS1123Label) of an annotation followed by a slash, second part for name segment of an annotation
+var IsValidKey = regexp.MustCompile(`^([a-z0-9]([-a-z0-9]*[a-z0-9])?\/)?[a-zA-Z0-9]([-a-zA-Z0-9_.]*[a-zA-Z0-9])?$`).MatchString
```
if we're going to do this validation, can we also add to the regexp that the length of the prefix must be {0,253} chars and length of the name must be {0,63} chars? This isn't a big deal and we can do without it but it's simple enough to add I think

adberger replied:
Please check the new regex carefully :)

natasha41575 commented on kustomize/commands/edit/set/setannotation.go:
```diff
@@ -0,0 +1,99 @@
+	}
+	for key := range m {
+		if IsValidKey(key) == false {
+			fmt.Println("Key is: " + key)
```
Please remove the print statement :)

adberger replied:
ugh very bad, thanks!

adberger commented:
/label tide/merge-method-squash

natasha41575 commented on kustomize/commands/edit/set/setannotation.go:
```diff
@@ -0,0 +1,99 @@
+	for key := range m {
+		if !IsValidKey(key) {
+			fmt.Println("Key is: " + key)
+			return errors.New("key doesn't match annotation rules fo keys")
```
Let's change the error message to something like:
`invalid annotation key: see the syntax and character set rules at https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/`

adberger commented:
@natasha41575 Do I need to do something in addition for another review from yours?

natasha41575 commented:
I think we need some more tests for the regex. For example, running `kustomize edit set annotation internal.config.kubernetes.io/generatorOptions:true` currently fails (as does anything else I try with a prefix). Please add tests for

1. The above example, `kustomize edit set annotation internal.config.kubernetes.io/generatorOptions:true`
2. An annotation that has a key with prefix 253 characters and a name with 63 characters.
3. An annotation that has a key with prefix 252 characters and a name with 62 characters.
4. An annotation that has a key with prefix 254 characters and a name with 64 characters.

adberger commented:
I fixed the regex as follows:
* Added `A-Z` for prefix
* Added `.` in prefix (but not start & end)
* Removed numerical characters `0-9` from start of prefix

I tried my best to extend the tests as you suggested, please check them and tell me if it's ok now.

natasha41575 commented on kustomize/commands/edit/set/setannotation_test.go:
```diff
@@ -0,0 +1,223 @@
+vwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz
+abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcde
+fghijklmnopqrstuvwxyzabcdefghijklmnopqrs/abcdefghijklmnopqrstuvwxyzabcdefghijklmnop
+qrstuvwxyzabcdefghijk`: "true"}
```
instead of using backticks (` `` `), let's do 

```
"abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstu" +
		"vwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz" +
		"abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcde" +
		"fghijklmnopqrstuvwxyzabcdefghijklmnopqrs/abcdefghijklmnopqrstuvwxyzabcdefghijklmnop" +
		"qrstuvwxyzabcdefghijk"
```

It seems a bit weird to have newlines in an annotation key (and this test also fails for me locally with backticks).

natasha41575 commented on kustomize/commands/edit/set/setannotation_test.go:
```diff
@@ -0,0 +1,223 @@
+jklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmn
+opqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrs
+tuvwxyzabcdefghijklmnopqrst/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabc
+defghij:true`}
```
same thing here, let's do 
```
"abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghi" +
		"jklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmn" +
		"opqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrs" +
		"tuvwxyzabcdefghijklmnopqrst/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabc" +
		"defghij:true"
```
instead of backticks to avoid the newlines.

natasha41575 commented on kustomize/commands/edit/set/setannotation_test.go:
```diff
@@ -0,0 +1,223 @@
+	err := cmd.RunE(cmd, args)
+	v.VerifyCall()
+	if err == nil {
+		t.Errorf("expected an error")
```
add `t.FailNow()` after this

adberger replied:
This seemed a bit odd to add this in only 2 places when there were 6 places which have `err==nil` condition.
Please let me know if all 6 need `t.FailNow()` or only the 2 you specified.
If only the 2, could you elaborate on why this is only needed there?

natasha41575 replied:
Yes, all 6 of them

natasha41575 commented on kustomize/commands/edit/set/setannotation_test.go:
```diff
@@ -0,0 +1,223 @@
+	err := cmd.RunE(cmd, args)
+	v.VerifyCall()
+	if err == nil {
+		t.Errorf("expected an error")
```
ditto, add `t.FailNow()` after this. This is to avoid panicking in line 124 if `err==nil`

natasha41575 reviewed (commented):
Looks pretty good! Just a couple of minor nits and we should be able to merge this soon. Thank you!

natasha41575 commented:
Tests look good; I'll pull the branch when I get some time to do some manual local testing as well.

natasha41575 commented:
This is almost there. Most cases such as `kustomize edit set internal.config.kubernetes.io/options: true` seem to be working as expected. 

But I find something very strange when running this locally:

```
$ kustomize edit set annotation abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkl/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijk:value
```

results in the following appearing in my kustomization file: 
```
commonAnnotations:
  ? abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkabcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijkl/abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcdefghijk
  : value
```

Any ideas why the question mark is showing up?

adberger commented:
@natasha41575 The "bug" not only exists for the `set annotation` command, try `set label` for example.
I tracked the issue down to the `JSONToYAML` function from the go yaml package (v1.2.0).

kf.Write(m)
-> data, err := mf.marshal(kustomization)
---> content, err := marshalField(field, kustomization)
-----> return yaml.Marshal(k)
-------->y, err := JSONToYAML(j)

https://imgur.com/a/4LiqB1v

`Println(t)` the output seems still correct
`Println(s)` the output is wrong

The problematic part seems to be somewhere in the `Marshal` function of the yaml.v2 package (v2.4.0):
```
func Marshal(in interface{}) (out []byte, err error) {
	defer handleErr(&err)
	e := newEncoder()
	defer e.destroy()
	e.marshalDoc("", reflect.ValueOf(in))
	e.finish()
	out = e.out
	return
}
```

My go knowledge isn't enough to really understand where the problem is, maybe someone can help me on this one.

natasha41575 commented:
> The "bug" not only exists for the set annotation command, try set label for example.
I tracked the issue down to the JSONToYAML function from the go yaml package (v1.2.0).

Weird. Since it also happens for `set label` I don't see a reason to hold this PR, but if you can, I would appreciate if you could file a follow-up issue with the same contents/explanation that you have in your above comment.

/lgtm
/approve
